app/models/loan.rb

```diff
@@ -189,4 +189,49 @@ def ensure_currency
   def amount_formatted
     ensure_currency.format_amount(amount)
   end
+
+  def signed_contract?
+    media.where(kind_value: 'contract').count > 0
+  end
+
+  def active?
+    status_value == 'active'
+  end
+
+  def incomplete_steps?
+    timeline_entries.merge(ProjectStep.past_due).count > 0
+  end
+
+  def sporadic_loan_updates?
+    timeline_entries.merge(ProjectStep.recent).count < [days_old, 30].min
+  end
+
+  def progress_pct
+    return 0 unless criteria
+    criteria.progress_pct
+  end
+
+  def health_warnings
+    warnings = []
+    warnings << :active_without_signed_contract if active? && !signed_contract?
+    warnings << :active_without_recent_logs if active? && most_recent_log_date < 30.days.ago
+    warnings << :past_due_steps if incomplete_steps?
+    warnings << :incomplete_loan_questions if progress_pct < 80
+    warnings << :sporadic_loan_updates if sporadic_loan_updates?
+    warnings
+  end
+
+  def healthy?
+    health_warnings.count < 1
+  end
+
+  def days_old
+    (end_date.beginning_of_day - signing_date.beginning_of_day) / (24 * 60 * 60)
+  end
+
+  private
+
+  def most_recent_log_date
+    project_logs.maximum(:date) || Time.at(0)
+  end
```

Comment:
The new methods in this file are nicely crafted, but I'm concerned about performance. Running the health checks for a single loan will kick off at least 5 non trivial queries:

* media
* timeline_entries x2
* progress_pct, which loads in the full LoanResponseSet and associated LoanQuestions
* project_logs

It's my understanding that this health check feature is designed to be run for a bunch of loans at once. Doing this will therefore kick of a ton of big queries and be super slow.